### PR TITLE
Automatically detect and skip loss spikes

### PR DESCRIPTION
**Pull request by Xirider**
**Patch Description**
Added a new flag `max_loss_to_skip_batch` that, if set to some maximum acceptable loss will abort the iteration before doing an optimizer step.
The loss value to compare to is the same one used in the logs. It might be or not different to the one in tensorboard.
The logic is similar to our `skip_gradient_update_on_clip_norm` flag which also skips batches, whenever the gradient norm is above the clip value, and also how we handle overflows.

**Testing steps**
Tested this with our small sweep script. I think our disks are full so I couldn't test this with a longer run.
For testing I increased the loss and checked whether we are skipping correctly

### CODE REVIEW

**Comment by Xirider** [comment]
Tested this and everything seems to work now!

I realized that testing this PR by randomly increasing loss with random.randint is not a good idea, as each gpu worker skips a different number of batches 🥲

**Comment by Xirider** [comment]
Also checked for changes in loss and other metrics. This is a 125m run:

1. With this PR and a loss spike every 10 steps:
image

2. with main:
image

Everything looks as expected.

File: metaseq/trainer.py [hunk 1]
```diff
@@ -966,6 +970,19 @@ def clip_grad_norm(
             skip_gradient_update_on_clip_norm=skip_gradient_update_on_clip_norm,
         )
 
+    def skip_spike(self, logging_outputs, max_loss_to_skip_batch):
+        if max_loss_to_skip_batch == -1:
+            return
+        loss_sum = sum(log.get("loss", 0) for log in logging_outputs)
+        sample_size = sum(log.get("sample_size", 0) for log in logging_outputs)
+        loss = float(loss_sum / sample_size / math.log(2))
```
Comment (Xirider) [inline metaseq/trainer.py:978]: This is the same calculation that happens for logging in reduce_metrics: https://github.com/facebookresearch/metaseq/blob/37809ec65632c434ed722a5c8aa37cd194ce9f8b/metaseq/criterions/cross_entropy.py#L103
The loss value used here matches the one that is logged.

File: metaseq/trainer.py [hunk 2]
```diff
@@ -809,6 +809,10 @@ def maybe_no_sync():
             )
             grad_norm = torch.tensor(0.0).cuda()
             self.zero_grad()
+        except SpikeError as e:
+            overflow = True
+            logger.info(str(e))
+            self.zero_grad()
```
Comment (Xirider) [inline metaseq/trainer.py:815]: This matches how overflow skips and gradient skips are handled

**Comment by Xirider** [comment]
> Just chatted with @stephenroller on this, and my main q here is: would it be possible to not hardcode a threshold for max loss but rather have it be defined as a function of some "moving average" of losses?
> 
> Ideally we would want to exclude "anomalies" which can be some function of some (exponential) moving average of previous losses.

Yes, I thought about this as well, but then we could have problems with spikes directly after a requeue if our ema is not filled at that point. Also it is more susceptible to weird training fluctuations.

A fixed value has the disadvantages that during the initial training restarts we will need to lower the value manually sometimes. But it seems for our current run, we won't need to do that again.

But I can modify the PR to skip at a multiple of EMA, if you think we want to be more automated :)

**Comment by Xirider** [comment]
This PR now uses a threshold value for the loss to loss ewm ratio to detect loss spikes, instead of a fixed loss value.

### Testing:
add this to sweep_baseline.py
```
hyperparam(
     "--ewm-ratio-to-skip-batch",
     1.25,
),
```

Result logs:
with a loss spike every 10 steps
[P564295826](https://www.internalfb.com/intern/paste/P564295826/)

with a loss spike every 10 steps and restarted from a previous checkpoint
[P564298095](https://www.internalfb.com/phabricator/paste/view/P564298095?lines=395)

File: metaseq/trainer.py [hunk 3]
```diff
@@ -966,6 +977,31 @@ def clip_grad_norm(
...
+        if ewm_ratio > ewm_ratio_to_skip_batch:
+            raise SpikeError(
+                f"Skip batch as we encountered a loss spike. In "
+                f"num_update: {self.get_num_updates()} the loss is {loss_t:.2f}. "
+                f"The ewm for the loss was only at {ewm_t:.2f} . "
+                f"The loss to ewm loss ratio is {ewm_ratio:.2f}, which is higher than "
+                f"ewm_ratio_to_skip_batch of {ewm_ratio_to_skip_batch} ."
+            )
+
+        self._ewm_loss = ewm_t
```
Comment (Xirider) [inline metaseq/trainer.py:1003]: Note: We are only updating the ewm if we keep the update. Otherwise the ewm would adapt quickly to bad batches and then stop skipping them.

File: metaseq/trainer.py [hunk 4]
```diff
@@ -1245,3 +1281,12 @@ def _set_module_by_path(module, path, value):
     for name in path[:-1]:
         module = getattr(module, name)
     setattr(module, path[-1], value)
+
+
+def ewm(loss_t, ewm_t_1, span):
+    alpha = 2 / (span + 1)
+    return (1 - alpha) * ewm_t_1 + alpha * loss_t
```
Comment (stephenroller) [inline metaseq/trainer.py:1288]: let's just inline this function please

File: metaseq/trainer.py (hunk 3 above)
Comment (stephenroller) [reply metaseq/trainer.py:1003]: Let's leave a comment explaining this please

File: metaseq/trainer.py [hunk 5]
```diff
@@ -966,6 +977,31 @@ def clip_grad_norm(
...
+            return
+        loss_sum = sum(log.get("loss", 0) for log in logging_outputs)
+        sample_size = sum(log.get("sample_size", 0) for log in logging_outputs)
+        loss_t = float(loss_sum / sample_size / math.log(2))
+
+        if self._ewm_loss is None:
+            self._ewm_loss = loss_t
+
+        ewm_t_1 = self._ewm_loss
+        ewm_t = ewm(loss_t, ewm_t_1, span=9)
```
Comment (stephenroller) [inline metaseq/trainer.py:991]: is it worth having ewm_loss logged to our metrics?

File: metaseq/trainer.py (hunk 2 above)
Comment (stephenroller) [reply metaseq/trainer.py:822]: I hope this works with skipping batches from pre-emption but i guess it should.

File: metaseq/trainer.py (hunk 5 above)
Comment (Xirider) [reply metaseq/trainer.py:991]: Might not be necessary as we can always calculate it again based on the logs? There is no use for the ewm_loss value outside of skipping. When a batch is skipped we are logging ewm_loss.

**Review by stephenroller — APPROVED** [review]
thanks, let's give it a try
